File: soln-backtracking-problems/subsets.java
/**
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/subsets/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> subsets=new ArrayList<>();
        generateSubsets(0,nums, new ArrayList<Integer>(),subsets);
        return subsets;
        
    }
    
    public void generateSubsets(int index, int[] nums, List<Integer> current, List<List<Integer>> subsets){
        subsets.add(new ArrayList<>(current));
        for(int i=index;i<nums.length;i++){
            current.add(nums[i]);
            generateSubsets(i+1,nums,current,subsets);
            current.remove(current.size()-1);
        }
    }
}

File: C++/soln-backtracking-problems/subsets.cpp
/**
 @author Farheen Bano
  
 Date 13-08-2021
  
 Reference-
 https://leetcode.com/problems/subsets/
*/

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> cur;
        generateSubsets(0,nums,cur,result);
        return result;
    }
    
    void generateSubsets(int index, vector<int>& nums, vector<int>& cur, vector<vector<int>>& result){        
        result.push_back(cur);

        for(int i=index;i<nums.size();i++){
            cur.push_back(nums[i]);
            generateSubsets(i+1,nums,cur,result);
            cur.pop_back();
        }
    }
};